ui/app/components/account-menu/account-menu.component.js

```diff
@@ -0,0 +1,301 @@
+    showAccountDetail: PropTypes.func,
+    showRemoveAccountConfirmationModal: PropTypes.func,
+    toggleAccountMenu: PropTypes.func,
+  }
+
+  state = {
+    atAccountListBottom: false,
+  }
+
+  componentDidUpdate (prevProps) {
+    const { prevIsAccountMenuOpen } = prevProps
+    const { isAccountMenuOpen } = this.props
+
+    if (!prevIsAccountMenuOpen && isAccountMenuOpen) {
+      this.setAtAccountListBottom()
+    }
+  }
+
+  renderAccounts () {
+    const {
+      identities,
+      accounts,
+      selectedAddress,
+      keyrings,
+      showAccountDetail,
+    } = this.props
+
+    const accountOrder = keyrings.reduce((list, keyring) => list.concat(keyring.accounts), [])
+
+    return accountOrder.filter(address => !!identities[address]).map(address => {
+      const identity = identities[address]
+      const isSelected = identity.address === selectedAddress
+
+      const balanceValue = accounts[address] ? accounts[address].balance : ''
+      const simpleAddress = identity.address.substring(2).toLowerCase()
+
+      const keyring = keyrings.find(kr => {
+        return kr.accounts.includes(simpleAddress) || kr.accounts.includes(identity.address)
+      })
+
+      return (
+        <div
+          className="account-menu__account menu__item--clickable"
+          onClick={() => showAccountDetail(identity.address)}
+          key={identity.address}
+        >
+          <div className="account-menu__check-mark">
+            { isSelected && <div className="account-menu__check-mark-icon" /> }
+          </div>
+          <Identicon
+            address={identity.address}
+            diameter={24}
+          />
+          <div className="account-menu__account-info">
+            <div className="account-menu__name">
+              { identity.name || '' }
+            </div>
+            <UserPreferencedCurrencyDisplay
+              className="account-menu__balance"
+              value={balanceValue}
+              type={PRIMARY}
+            />
+          </div>
+          { this.renderKeyringType(keyring) }
+          { this.renderRemoveAccount(keyring, identity) }
+        </div>
+      )
+    })
+  }
+
+  renderRemoveAccount (keyring, identity) {
+    const { t } = this.context
+    // Any account that's not from the HD wallet Keyring can be removed
+    const { type } = keyring
+    const isRemovable = type !== 'HD Key Tree'
+
+    return isRemovable && (
+      <Tooltip
+        title={t('removeAccount')}
+        position="bottom"
+      >
+        <a
+          className="remove-account-icon"
+          onClick={e => this.removeAccount(e, identity)}
+        />
+      </Tooltip>
+    )
+  }
+
+  removeAccount (e, identity) {
+    e.preventDefault()
+    e.stopPropagation()
+    const { showRemoveAccountConfirmationModal } = this.props
+    showRemoveAccountConfirmationModal(identity)
+  }
+
+  renderKeyringType (keyring) {
+    const { t } = this.context
+
+    try { // Sometimes keyrings aren't loaded yet:
```

Comment:
In the case where keyrings aren't loaded, what exception are we catching here? Can we instead have a default for `type` or `return null` if `!keyring.type`?